File: frontend/src/types/graph.ts
import type { FlowScope } from './flow'

export type GraphNodeType = 'vm' | 'unknown_internal' | 'external' | 'unknown'

export interface GraphNode {
  id: string
  type: GraphNodeType
  label: string
  ip?: string
  private_ip?: string
  public_ip?: string
  status?: string
  tenant_id?: string
  role?: string
  traffic_in: number
  traffic_out: number
}

export interface GraphEdge {
  id: string
  source: string
  target: string
  protocol: 'tcp' | 'udp' | string
  dst_port: number
  scope: FlowScope
  bytes_sent: number
  bytes_received: number
  packets: number
  connection_count: number
  request_count: number
  retransmission_count?: number
  first_seen: string
  last_seen: string
  last_observed_at: string
  active: boolean
  active_until: string
  failed?: boolean
  failed_until?: string
  weight: number
  kind?: 'traffic' | 'reachability' | string
  reachable?: boolean
  source_ip?: string
  dest_ip?: string
  source_role?: string
  dest_role?: string
  direction?: string
  server_port?: number
  error_count?: number
  last_error_at?: string
  total_bytes?: number
  avg_rtt_ms?: number
  avg_app_delay_ms?: number
  http_2xx_count?: number
  http_4xx_count?: number
  http_5xx_count?: number
  last_http_status?: number
  p95_rtt_ms?: number
  avg_response_duration_ms?: number
  last_response_code?: number
  agent_ids?: string[]
  observation_points?: string[]
}

export type ConnectionHealth = 'healthy' | 'degraded' | 'failed' | 'inactive' | 'unknown'

export interface ConnectionSummary {
  id: string
  source: string
  target: string
  source_label: string
  target_label: string
  source_ip?: string
  target_ip?: string
  direction: 'one_way' | 'two_way'
  active_direction: 'source_to_target' | 'target_to_source' | 'bidirectional' | 'none'
  health: ConnectionHealth
  scope: string
  protocols: string[]
  ports: number[]
  connected: boolean
  active: boolean
  failed: boolean
  slow: boolean
  request_count: number
  error_count: number
  retransmission_count?: number
  total_bytes: number
  packets: number
  connection_count: number
  avg_rtt_ms: number
  p95_rtt_ms: number
  avg_response_duration_ms: number
  last_response_code?: number
  last_seen?: string
  last_observed_at?: string
  agent_ids: string[]
  observation_points: string[]
  recommendation: string
}

export interface GraphData {
  nodes: GraphNode[]
  edges: GraphEdge[]
}

export interface GraphFilters {
  vm_id: string
  scope: string
  protocol: string
  port: string
  time_range: string
  min_bytes: string
  status: string
}

File: backend/internal/model/graph.go
package model

import "time"

type Graph struct {
	Nodes []GraphNode `json:"nodes"`
	Edges []GraphEdge `json:"edges"`
}

type GraphNode struct {
	ID         string `json:"id"`
	Type       string `json:"type"`
	Label      string `json:"label"`
	IP         string `json:"ip,omitempty"`
	Status     string `json:"status,omitempty"`
	TenantID   string `json:"tenant_id,omitempty"`
	Role       string `json:"role,omitempty"`
	TrafficIn  int64  `json:"traffic_in"`
	TrafficOut int64  `json:"traffic_out"`
}

type GraphEdge struct {
	ID              string     `json:"id"`
	Source          string     `json:"source"`
	Target          string     `json:"target"`
	Protocol        string     `json:"protocol"`
	DstPort         int        `json:"dst_port"`
	Scope           string     `json:"scope"`
	BytesSent       int64      `json:"bytes_sent"`
	BytesReceived   int64      `json:"bytes_received"`
	Packets         int64      `json:"packets"`
	ConnectionCount int64      `json:"connection_count"`
	RequestCount    int64      `json:"request_count"`
	ErrorCount      int64      `json:"error_count"`
	FirstSeen       time.Time  `json:"first_seen"`
	LastSeen        time.Time  `json:"last_seen"`
	LastObservedAt  time.Time  `json:"last_observed_at"`
	Active          bool       `json:"active"`
	ActiveUntil     time.Time  `json:"active_until"`
	Failed          bool       `json:"failed"`
	FailedUntil     time.Time  `json:"failed_until,omitempty"`
	LastErrorAt     *time.Time `json:"last_error_at,omitempty"`
	Weight          int        `json:"weight"`
	Kind            string     `json:"kind,omitempty"`
	Reachable       bool       `json:"reachable,omitempty"`
	AvgRTTMs        float64    `json:"avg_rtt_ms,omitempty"`
}

type GraphFilter struct {
	AgentID   string
	TenantID  string
	VMID      string
	Scope     string
	Protocol  string
	Port      int
	TimeRange time.Duration
	MinBytes  int64
	Status    string
}
